src/server/raftkv.rs

```diff
@@ -208,6 +198,95 @@ pub fn drop_snapshot_callback<T>() -> kv::Result<T> {
     Err(kv::Error::from(kv::ErrorInner::Request(err)))
 }
 
+struct WriteResCore {
+    ev: AtomicU8,
+    result: UnsafeCell<Option<kv::Result<()>>>,
+    wake: AtomicWaker,
+}
+
+struct WriteResSub {
+    notified_ev: u8,
+    core: Arc<WriteResCore>,
+}
+
+unsafe impl Send for WriteResSub {}
+
+impl Stream for WriteResSub {
+    type Item = WriteEvent;
+
+    #[inline]
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let mut s = self.as_mut();
+        let mut cur_ev = s.core.ev.load(Ordering::Acquire);
+        if cur_ev == s.notified_ev {
+            s.core.wake.register(cx.waker());
+            cur_ev = s.core.ev.load(Ordering::Acquire);
+            if cur_ev == s.notified_ev {
+                return Poll::Pending;
+            }
+        }
+        s.notified_ev = cur_ev;
+        match cur_ev {
```

Comment:
This means a later event will overwrite the previous one even if the event is not processed?

Then, if a request subscribes `Proposed` but `core.ev` already becomes `Committed` when it polls, it will  miss the `Proposed` event.

Reply:
Yes, it's possible generally. But it won't happen for any known usage. Scheduler will either subscribe proposed or committed not both. If the event is overwritten, it must be written by final result.